contracts/utils/StaticDelegateCall.sol

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+/**
+ * @dev Provides emulation of static delegatecalls for implementing view-only
+ * functions that execute another contract's code with the calling contract's
+ * context.
+ */
+abstract contract StaticDelegateCall {
+    /**
+     * @dev Perform a static delegate call to the specified `target` with
+     * calldata `data`. Returns success boolean and return data.
+     */
+    function _staticDelegateCall(address target, bytes memory data) internal view returns (bool, bytes memory) {
+        /* This staticcall always reverts, so we can ignore the success boolean */
+        (, bytes memory returndata) = address(this).staticcall(
+            abi.encodeCall(this.delegateCallAndRevert, (target, data))
+        );
+        return abi.decode(returndata, (bool, bytes));
+    }
+
+    /**
+     * @dev Simulate a delegatecall to the specified `target` with calldata
+     * `data` by reverting with the results. The resulting success boolean and
+     * return data are ABI-encoded in the revert data.
+     *
+     * Inspired by simulateAndRevert() from Gnosis' StorageSimulation contract.
+     */
+    function delegateCallAndRevert(address target, bytes memory data) external {
+        require(msg.sender == address(this), "StaticDelegateCall: caller must be self");
```

Comment:
Out of curiosity, what is the attack/issue that this require protect against?

Reply:
In this case, it's to ensure that the feature doesn't introduce any unexpected side effects that the user didn't opt into. Without the `msg.sender` check, `delegateCallAndRevert()` exposes all of the contract's storage to any other contract onchain, which isn't necessarily a security issue (in many ways it's a feature), but is a substantial side effect that didn't exist prior to inheriting the contract. If the user does want that effect, they would opt into knowingly by creating an external wrapper for `_staticDelegateCall()` that explicitly exposes it to other contracts.